File: Matlab/read_cnf.m
%Function that takes in cnf file and return a matrix of Clauses C and the
%number of variables for the given UNSAT problem
 
function [Clauses,var,C]= read_cnf(cnffile,k_sat)
    
    C=0; % The number of clauses
    m=k_sat; % k-SAT problem 
    var=0; %The number of variables
    Clauses =0; % Storing the values for each clause in a matrix
    
    fid = fopen(cnffile);
    if fid<0
        return
    end
    k=1;
    while ~feof(fid)
        tline = fgetl(fid);
        tline = strtrim(tline);
        if isempty(tline)
            continue;
        end
        if (lower(tline(1))=='p')
            X = (string(split(tline)));
            var = str2num(X(end -1));
            C= str2num(X(end));
            Clauses = zeros(C,m); 
            continue;
        end
        
        if (tline(1)~= 'c') &&(k<=C)
            x = str2num(tline);
            Clauses(k,:)= x(1:end-1);
            k=k+1;
        end
    end
    fclose(fid);
end

    

File: Python/kSAT_to_3SAT_solver/read_cnf.py
import numpy as np

def read_cnf(cnffile, k_sat):
    
    NC=0
    m=k_sat 
    var=0 
    Clauses =[]
    k=1
    with open(cnffile, "r") as fid:
    
        for line in fid:
            tline = line.strip()
            if (tline.lower().startswith('p')) :
                X = tline.split()
                var = int(X[-2])
                varo = var
                NC = int(X[-1])
                continue;
                
            if (not tline.lower().startswith('c'))  and  (k<= NC) :
                x = np.fromstring(tline,dtype = int, sep=' ')
                x = x.tolist()
                Clauses.append(x[:-1])
                k=k+1

    return Clauses,var,NC